simulation/halsim_lowfi/src/main/native/cpp/HALSimLowFi.cpp

```diff
@@ -0,0 +1,60 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2017 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "HALSimLowFi.h"
+
+void HALSimLowFi::Initialize() {
+  table = nt::NetworkTableInstance::GetDefault().GetTable("sim");
+}
+
+void HALSimNTProvider::Inject(std::shared_ptr<HALSimLowFi> parentArg,
+                              std::string tableNameArg) {
+  parent = parentArg;
+  tableName = tableNameArg;
+  table = parent->table->GetSubTable(tableName);
+
+  this->Initialize();
+}
+
+void NTProviderBaseCallback(const char* name, void* param,
+                            const struct HAL_Value* value) {
+  auto info =
+      static_cast<struct HALSimNTProvider::NTProviderCallbackInfo*>(param);
+  uint32_t chan = static_cast<uint32_t>(info->channel);
+  auto provider = info->provider;
+  auto table = info->table;
+  provider->OnCallback(chan, table);
+}
+
+void HALSimNTProvider::InitializeDefault(
+    int numChannels, HALCbRegisterIndexedFunc registerFunc) {
+  this->numChannels = numChannels;
+  cbInfos.reserve(numChannels);
+  for (int i = 0; i < numChannels; i++) {
+    struct NTProviderCallbackInfo info = {
+        this, table->GetSubTable(tableName + std::to_string(i)), i};
```

Comment:
I think this can be `tableName + llvm::Twine(I)` to avoid multiple allocations. And maybe we should be using emplace back with an inline initialization.

Reply:
I've added llvm::Twine and emplace_back but I'm keeping the struct initialization on the line above since inline initialization of a struct in a func call is not C++ standard.